staging/src/k8s.io/apiserver/pkg/endpoints/request/context_shard.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"context"
+)
+
+type shardKey int
+
+const (
+	// shardKey is the context key for the request.
+	shardContextKey shardKey = iota
+)
+
+// Shard describes a shard
+type Shard string
+
+// Empty returns true if the name of the shard is empty.
+func (s Shard) Empty() bool {
+	return s == ""
+}
+
+// Wildcard checks if the given shard name matches wildcard.
+// If true the query applies to all shards.
+func (s Shard) Wildcard() bool {
+	return s == "*"
+}
+
+// String casts Shard to string type
+func (s Shard) String() string {
+	return string(s)
+}
+
+// WithShard returns a context that holds the given shard.
+func WithShard(parent context.Context, shard Shard) context.Context {
+	return context.WithValue(parent, shardContextKey, shard)
+}
+
+// ShardFrom returns the value of the shard key in the context, or an empty value if there is no shard key.
+func ShardFrom(ctx context.Context) Shard {
+	shard, ok := ctx.Value(shardContextKey).(Shard)
```

Comment:
it could, it will just panic if the value under the key cannot be cast to `Shard`

Reply:
The `shardContextKey` is private to this package. The only way that it would panic is if someone wrote code _in this package_ to add the wrong thing. /shrug

Reply:
yes, I just don't like when the code panics but sure I can change it.

Reply:
and it broke the kcp server
```
I0901 10:30:42.331404   38583 etcd3.go:306] Start monitoring storage db size metric for endpoint https://localhost:45743 with polling interval 30s
        panic: interface conversion: interface {} is nil, not request.Shard
        
        goroutine 1 [running]:
        k8s.io/apiserver/pkg/endpoints/request.ShardFrom(...)
        	/go/pkg/mod/github.com/p0lyn0mial/kubernetes/staging/src/k8s.io/apiserver@v0.0.0-20220901092652-1cf194952b52/pkg/endpoints/request/context_shard.go:56
 ```